framework/src/materials/CoupledValueFunctionMaterial.C

```diff
@@ -0,0 +1,90 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "CoupledValueFunctionMaterial.h"
+#include "Function.h"
+
+registerMooseObject("MooseApp", CoupledValueFunctionMaterial);
+registerMooseObject("MooseApp", ADCoupledValueFunctionMaterial);
+
+template <bool is_ad>
+InputParameters
+CoupledValueFunctionMaterialTempl<is_ad>::validParams()
+{
+  InputParameters params = Material::validParams();
+  params.addClassDescription("Compute a function value from coupled variables");
+  params.addRequiredParam<FunctionName>("function",
+                                        "Coupled function to evaluate with values from v");
+  MultiMooseEnum parameter_order_enum("X Y Z T");
+  params.addParam<MultiMooseEnum>("parameter_order",
+                                  parameter_order_enum,
+                                  "If provided, an entry per couple variable specifies "
+                                  "the function argument it should apply to.");
+  params.addRequiredParam<MaterialPropertyName>("prop_name", "Output property name");
+  params.addCoupledVar(
+      "v",
+      "List of up to four coupled variables that are substituted for x,y,z, and t "
+      "in the coupled function (or the order chosen using the `parameter_order` parameter)");
+  return params;
+}
+
+template <bool is_ad>
+CoupledValueFunctionMaterialTempl<is_ad>::CoupledValueFunctionMaterialTempl(
+    const InputParameters & parameters)
+  : Material(parameters),
+    _prop(declareGenericProperty<Real, is_ad>(getParam<MaterialPropertyName>("prop_name"))),
+    _function(getFunction("function")),
+    _vals(coupledGenericValues<is_ad>("v")),
+    _nvals(coupledComponents("v"))
```

Comment:
Where do these variables need to live? Would this work for boundary restricted variables?

Reply:
This is not different from any other material. So I suppose yes.